README
= Description
  This is a drop-in replacement for the find module currently in the standard
  library. It is modeled on a typical 'find' command found on most Unix systems.

= Installation

  gem install file-find

= Synopsis
  require 'file/find'

  rule = File::Find.new(
    :pattern => "*.rb",
    :follow  => false,
    :path    => ['/usr/local/lib', '/opt/local/lib']
  )

  rule.find{ |f|
    puts f
  }

= Rationale
  The current find module in the standard library is inadequate. It is, quite
  frankly, not much more than a plain Dir.glob call. This library provides an
  interface based on options typically available on your command line 'find'
  command, thus allowing you much greater control over how you find your files.

= Options
  * atime
  * ctime
  * follow
  * ftype
  * inum
  * group (name or id)
  * maxdepth
  * mindepth
  * mount
  * mtime
  * name (or 'pattern')
  * path
  * perm
  * prune
  * size
  * user (name or id)

  In addition to the above options, FileTest methods such as 'readable?' and
  'writable?' may be used as keys, with true or false for their values.

  See the RDoc documentation for more details about these options.

= Future Plans
  None at this time. Please log any feature requests on the project page at:

  http://github.com/djberg96/file-find

= Options I won't support
  Generally speaking, anything that would require mucking around with C code
  or is just too difficult to implement in a cross platform manner will not be
  supported. These include the following options:

  * acl/xattr - Way too difficult to implement in a cross platform manner, and
    a rarely used option in practice.

  * cpio/ncpio - I will not shell out to this or any other 3rd party
    application.

  * ls/print - Use Ruby's builtin printing methods to print as you see fit.

  * ok - This is not interactive software.

= Known Issues
  The :perm option is limited to 0644 and 0444 on MS Windows.

  The :user, :group, and :inum options require the win32-file gem to work
  properly on MS Windows. However, it is not officially a dependency.

= Bugs
  None that I'm aware of beyond the ones mentioned in the Known Issues. Please
  log any bug reports on the project page at:

  http://github.com/djberg96/file-find

= Acknowledgements
  * Richard Clamp's File::Find::Rule Perl module for additional ideas and
    inspiration.
  * Bill Kleb for ideas regarding name, group and perm enhancements.
  * Hal Fulton for his implementation of symbolic permissions.

= License
  Apache-2.0

= Copyright
  (C) 2007-2019, Daniel J. Berger, All Rights Reserved

= Author
  Daniel J. Berger


CHANGES
== 0.4.2 - 6-May-2020
* Added LICENSE file as required by the Apache-2.0 license.

== 0.4.1 - 23-Jan-2019
* Added metadata to the gemspec.

== 0.4.0 - 26-Dec-2018
* License changed to Apache-2.0.
* Fixed a bug where mindepth and maxdepth would break if more than one
  path was specified as part of the rule. Thanks go to flagos for the spot.
* The VERSION constant is now frozen.
* The cert has been updated.

== 0.3.9 - 16-Dec-2015
* This gem is now signed.
* Added a file-find.rb file for convenience.
* Updates to the Rakefile and gemspec.

== 0.3.8 - 12-Feb-2014
* Removed all references to the Etc module since the sys-admin library now uses
  FFI and works with JRuby.
* The :user, :group, and :inum options will now work on MS Windows if you have
  the win32-file gem installed.
* Fixed the perm option on Windows for its limited subset of available perms.
* You can now include a leading zero for the perm option if you wish.
* Some tests that were previously omitted on Windows are now included.
* Updates to the test suite, readme, etc.

== 0.3.7 - 15-Jan-2014
* Fixed a bug with brackets in the path name. Thanks go to Jeremy Lawler
  for the spot and the patch.

== 0.3.6 - 6-Sep-2013
* Removed rbconfig from library and test file. Just use File::ALT_SEPARATOR
  to check for Windows instead.
* Removed unused variables.
* Updated dev dependencies.

== 0.3.5 - 15-Jul-2011
* Fixed a bug with the :follow option.
* Gemspec, Rakefile and test cleanup.

== 0.3.4 - 19-Sep-2009
* Fixed a packaging bug. Thanks go to Gabriel Horner for the spot.
* Added the 'gem' task to the Rakefile for building the gem. Removed the
  gem builder code that was in the gemspec itself.
* Updated the dependency for sys-admin to 1.5.2.

== 0.3.3 - 3-Aug-2009
* Now compatible with Ruby 1.9.x.
* Added support for the :links option
* Updated the :mount accessor to work properly if the value is changed
  after the File::Find object is initially created.
* Eliminated some 'shadowed variable' warnings in Ruby 1.9.x.
* Added a fuller description to the gemspec.
* Minor test refactoring.

== 0.3.2 - 20-Feb-2009
* Added support for the :mount option.
* Added some basic tests for the :mount option.

== 0.3.1 - 9-Jan-2009
* Now defaults to using Etc instead of Sys::Admin for implementations
  that don't support building C extensions, e.g. JRuby.
* Updated the test suite to work with JRuby.
* Minor optimizations for symbolic perms and atime/ctime/mtime checks.
  Thanks go in part to Ryan Davis' flay library.

== 0.3.0 - 30-Dec-2008
* Added support for FileTest operations. All options passed to the constructor
  that end with '?' are now validated and treated as FileTest operations.

== 0.2.5 - 29-Dec-2008
* Added maxdepth and mindepth support.
* Added a 'clean' rake task to cleanup any test-unit results files.

== 0.2.4 - 10-Dec-2008
* Added support for symbolic permissions. Thanks go to Bill Kleb for the
  suggestion and to Hal Fulton for providing the solution.

== 0.2.3 - 25-Nov-2008
* Added mtime support. My previous concerns, which I believe stemmed from
  the find(2) man page on Solaris 10 with regards to atime checks modifying
  the mtime, appear have been unfounded.

== 0.2.2 - 19-Nov-2008
* The :user and :group options now accept a name or a numeric id. Thanks go
  to Bill Kleb for the suggestion.
* Fixed yet another path bug for MS Windows.
* Updated platform check to use CONFIG instead of RUBY_PLATFORM, because the
  latter does not work as I intended with other implementations, e.g. JRuby.
* Added sys-admin and test-unit as prerequisites.
* Added tests for the :user and :group options.

== 0.2.1 - 4-Oct-2007
* Added the File::Find#previous method, which lets you see the previous
  match, if any.
* Path name bug fix for MS Windows.
* Test suite bug fixes for MS Windows (perm test now skipped).
* Inaccessible directories are now skipped instead of raising an error.

== 0.2.0 - 26-Apr-2007
* Fixed a bug where it was not traversing subdirectories.
* Added support for the perm and prune options.

== 0.1.1 - 25-Apr-2007
* The default for name is now '*', i.e. everything.
* Fixed a bug where directories were not matched. Thanks go to Leslie Viljoen
  for the spot.
* The size option now accepts strings with comparable operators. For example,
  you can now look for files greater than 400 bytes with the string "> 400".

== 0.1.0 - 24-Apr-2007
* Initial release


lib/file/find.rb
require 'date'
require 'sys/admin'

begin
  require 'win32/file'
rescue LoadError
  # Do nothing, not required, just nicer.
end

class File::Find
  # The version of the file-find library
  VERSION = '0.4.2'.freeze

  # :stopdoc:
  VALID_OPTIONS = %w[
    atime
    ctime
    follow
    ftype
    inum
    group
    links
    maxdepth
    mindepth
    mount
    mtime
    name
    pattern
    path
    perm
    prune
    size
    user
  ]
  # :startdoc:

  # The starting path(s) for the search. The default is the current directory.
  # This can be a single path or an array of paths.
  #
  attr_accessor :path

  # The list of options passed to the constructor and/or used by the
  # File::Find#find method.
  #
  attr_accessor :options

  # Limits searches by file access time, where the value you supply is the
  # number of days back from the time that the File::Find#find method was
  # called.
  #
  attr_accessor :atime

  # Limits searches by file change time, where the value you supply is the
  # number of days back from the time that the File::Find#find method was
  # called.
  #
  attr_accessor :ctime

  # Limits searches to files that belong to a specific group, where the
  # group can be either a group name or ID.
  #
  attr_accessor :group

  # An array of two element arrays for storing FileTest methods and their
  # boolean value.
  #
  attr_accessor :filetest

  # Controls the behavior of how symlinks are followed. If set to true (the
  # default), then follows the file pointed to. If false, it considers the
  # symlink itself.
  #
  attr_accessor :follow

  # Limits searches to specific types of files. The possible values here are
  # those returned by the File.ftype method.
  #
  attr_accessor :ftype

  # Limits search to a file with a specific inode number.
  #
  attr_accessor :inum

  # Limits search to files with the specified number of links.
  #
  attr_accessor :links

  # Limits search to a maximum depth into the tree relative to the starting
  # search directory.
  #
  attr_accessor :maxdepth

  # Limits searches to a minimum depth into the tree relative to the starting
  # search directory.
  #
  attr_accessor :mindepth

  # Limits searches to the same filesystem as the specified directory. For
  # Windows users, this refers to the volume.
  #
  attr_reader :mount

  # Limits searches by file modification time, where the value you supply is
  # the number of days back from the time that the File::Find#find method was
  # called.
  #
  attr_accessor :mtime

  # The name pattern used to limit file searches. The patterns that are legal
  # for Dir.glob are legal here. The default is '*', i.e. everything.
  #
  attr_accessor :name

  # Limits searches to files which have permissions that match the octal
  # value that you provide. For purposes of this comparison, only the user,
  # group, and world settings are used.
  #
  # You may optionally use symbolic permissions, e.g. "g+rw", "u=rwx", etc.
  #
  # MS Windows only recognizes two modes, 0644 and 0444.
  #
  attr_accessor :perm

  # Skips files or directories that match the string provided as an argument.
  #
  attr_accessor :prune

  # If the value passed is an integer, this option limits searches to files
  # that match the size, in bytes, exactly. If a string is passed, you can
  # use the standard comparable operators to match files, e.g. ">= 200" would
  # limit searches to files greater than or equal to 200 bytes.
  #
  attr_accessor :size

  # Limits searches to files that belong to a specific user, where the user
  # can be either a user name or an ID.
  #
  attr_accessor :user

  # The file that matched previously in the current search.
  #
  attr_reader :previous

  alias pattern name
  alias pattern= name=

  # Creates and returns a new File::Find object. The options set for this
  # object serve as the rules for determining what files the File::Find#find
  # method will search for.
  #
  # In addition to the standard list of valid options, you may also use
  # FileTest methods as options, setting their value to true or false.
  #
  # Example:
  #
  #    rule = File::Find.new(
  #       :name      => "*.rb",
  #       :follow    => false,
  #       :path      => ['/usr/local/lib', '/opt/local/lib'],
  #       :readable? => true
  #    )
  #
  def initialize(options = {})
    @options = options

    @atime  = nil
    @ctime  = nil
    @ftype  = nil
    @group  = nil
    @follow = true
    @inum   = nil
    @links  = nil
    @mount  = nil
    @mtime  = nil
    @perm   = nil
    @prune  = nil
    @size   = nil
    @user   = nil

    @previous = nil
    @maxdepth = nil
    @mindepth = nil
    @filetest = []

    validate_and_set_options(options) unless options.empty?

    @filesystem = File.stat(@mount).dev if @mount

    @path ||= Dir.pwd
    @name ||= '*'
  end

  # Executes the find based on the rules you set for the File::Find object.
  # In block form, yields each file in turn that matches the specified rules.
  # In non-block form it will return an array of matches instead.
  #
  # Example:
  #
  #   rule = File::Find.new(
  #      :name    => "*.rb",
  #      :follow  => false,
  #      :path    => ['/usr/local/lib', '/opt/local/lib']
  #   )
  #
  #   rule.find{ |f|
  #      puts f
  #   }
  #
  def find
    results = [] unless block_given?
    paths   = @path.is_a?(String) ? [@path] : @path # Ruby 1.9.x compatibility

    if @prune
      prune_regex = Regexp.new(@prune)
    else
      prune_regex = nil
    end

    paths.each{ |path|
      begin
        Dir.foreach(path){ |file|
          next if file == '.'
          next if file == '..'

          if prune_regex
            next if prune_regex.match(file)
          end

          file = File.join(path, file)

          stat_method = @follow ? :stat : :lstat
          # Skip files we cannot access, stale links, etc.
          begin
            stat_info = File.send(stat_method, file)
          rescue Errno::ENOENT, Errno::EACCES
            next
          rescue Errno::ELOOP
            stat_method = :lstat # Handle recursive symlinks
            retry if stat_method.to_s != 'lstat'
          end

          # We need to escape any brackets in the directory name.
          glob = File.join(File.dirname(file).gsub(/([\[\]])/,'\\\\\1'), @name)

          # Dir[] doesn't like backslashes
          if File::ALT_SEPARATOR
            file.tr!(File::ALT_SEPARATOR, File::SEPARATOR)
            glob.tr!(File::ALT_SEPARATOR, File::SEPARATOR)
          end

          if @mount
            next unless stat_info.dev == @filesystem
          end

          if @links
            next unless stat_info.nlink == @links
          end

          if @maxdepth || @mindepth
            file_depth = file.split(File::SEPARATOR).length
            current_base_path = [@path].flatten.find{ |tpath| file.include?(tpath) }
            path_depth = current_base_path.split(File::SEPARATOR).length

            depth = file_depth - path_depth

            if @maxdepth && (depth > @maxdepth)
              if File.directory?(file)
                unless paths.include?(file) && depth > @maxdepth
                  paths << file
                end
              end

              next
            end

            if @mindepth && (depth < @mindepth)
              if File.directory?(file)
                unless paths.include?(file) && depth < @mindepth
                  paths << file
                end
              end

              next
            end
          end

          # Add directories back onto the list of paths to search unless
          # they've already been added
          #
          if stat_info.directory?
            paths << file unless paths.include?(file)
          end

          next unless Dir[glob].include?(file)

          unless @filetest.empty?
            file_test = true

            @filetest.each{ |array|
              meth = array[0]
              bool = array[1]

              unless File.send(meth, file) == bool
                file_test = false
                break
              end
            }

            next unless file_test
          end

          if @atime || @ctime || @mtime
            date1 = Date.parse(Time.now.to_s)

            if @atime
              date2 = Date.parse(stat_info.atime.to_s)
              next unless (date1 - date2).numerator == @atime
            end

            if @ctime
              date2 = Date.parse(stat_info.ctime.to_s)
              next unless (date1 - date2).numerator == @ctime
            end

            if @mtime
              date2 = Date.parse(stat_info.mtime.to_s)
              next unless (date1 - date2).numerator == @mtime
            end
          end

          if @ftype
            next unless File.ftype(file) == @ftype
          end

          if @group
            if @group.is_a?(String)
              if File::ALT_SEPARATOR
                begin
                  next unless Sys::Admin.get_group(stat_info.gid, :LocalAccount => true).name == @group
                rescue Sys::Admin::Error
                  next
                end
              else
                begin
                  next unless Sys::Admin.get_group(stat_info.gid).name == @group
                rescue Sys::Admin::Error
                  next
                end
              end
            else
              next unless stat_info.gid == @group
            end
          end

          if @inum
            next unless stat_info.ino == @inum
          end

          # Note that only 0644 and 0444 are supported on MS Windows.
          if @perm
            if @perm.is_a?(String)
              octal_perm = sym2oct(@perm)
              next unless stat_info.mode & octal_perm == octal_perm
            else
              next unless sprintf("%o", stat_info.mode & 07777) == sprintf("%o", @perm)
            end
          end

          # Allow plain numbers, or strings for comparison operators.
          if @size
            if @size.is_a?(String)
              regex = /^([><=]+)\s*?(\d+)$/
              match = regex.match(@size)

              if match.nil? || match.captures.include?(nil)
                raise ArgumentError, "invalid size string: '#{@size}'"
              end

              operator = match.captures.first.strip
              number   = match.captures.last.strip.to_i

              next unless stat_info.size.send(operator, number)
            else
              next unless stat_info.size == @size
            end
          end

          if @user
            if @user.is_a?(String)
              if File::ALT_SEPARATOR
                begin
                  next unless Sys::Admin.get_user(stat_info.uid, :LocalAccount => true).name == @user
                rescue Sys::Admin::Error
                  next
                end
              else
                begin
                  next unless Sys::Admin.get_user(stat_info.uid).name == @user
                rescue Sys::Admin::Error
                  next
                end
              end
            else
              next unless stat_info.uid == @user
            end
          end

          if block_given?
            yield file
          else
            results << file
          end

          @previous = file unless @previous == file
        }
      rescue Errno::EACCES
        next # Skip inaccessible directories
      end
    }

    block_given? ? nil : results
  end

  # Limits searches to the same file system as the specified +mount_point+.
  #
  def mount=(mount_point)
    @mount = mount_point
    @filesystem = File.stat(mount_point).dev
  end

  private

  # This validates that the keys are valid. If they are, it sets the value
  # of that key's corresponding method to the given value. If a key ends
  # with a '?', it's validated as a File method.
  #
  def validate_and_set_options(options)
    options.each do |key, value|
      key = key.to_s.downcase

      if key[-1].chr == '?'
        sym = key.to_sym

        unless File.respond_to?(sym)
          raise ArgumentError, "invalid option '#{key}'"
        end

        @filetest << [sym, value]
      else
        unless VALID_OPTIONS.include?(key)
          raise ArgumentError, "invalid option '#{key}'"
        end

        send("#{key}=", value)
      end
    end
  end

  # Converts a symoblic permissions mode into its octal equivalent.
  #--
  # Taken almost entirely from ruby-talk: 96956 (Hal Fulton).
  #
  def sym2oct(str)
    left  = {'u' => 0700, 'g' => 0070, 'o' => 0007, 'a' => 0777}
    right = {'r' => 0444, 'w' => 0222, 'x' => 0111}
    regex = /([ugoa]+)([+-=])([rwx]+)/

    cmds = str.split(',')

    perm = 0

    cmds.each do |cmd|
      match = cmd.match(regex)
      raise "Invalid symbolic permissions: '#{str}'" if match.nil?

      who, what, how = match.to_a[1..-1]

      who  = who.split(//).inject(0){ |num,b| num |= left[b]; num }
      how  = how.split(//).inject(0){ |num,b| num |= right[b]; num }
      mask = who & how

      case what
        when '+'
          perm = perm | mask
        when '-'
          perm = perm & ~mask
        when '='
          perm = mask
      end
    end

    perm
  end
end
